### DISCUSSION TITLE
Luau vectors

### DISCUSSION THREAD

**Discussion by setzer22** [mlua-rs/mlua#140] [Q&A] [answered]
I'm testing the new Luau support for native vectors in `0.8.0-beta.2`. I can see the `mlua::Value` vector type and I've succesfully been able to pass a vector from Rust to Lua like so:

```rust
let lua = Lua::new();
let x: mlua::Function = lua
    .load("return function(v) print(v) end")
    .call(())
    .unwrap();
x.call::<_, ()>(mlua::Value::Vector(1.0, 2.0, 3.0));
```

But I can't create native vectors on the Lua side. The documentation suggests it's `vector(1,2,3)`, but it doesn't seem to be working. Calling `return vector(1,2,3)` in Lua code tells me I'm trying to call a `nil` table (vector).

**Comment by khvzak** [comment, marked as answer]
The constructor indeed is not provided by Luau, you need to do something like:
```rust
    let globals = lua.globals();
    globals.set(
        "vector",
        lua.create_function(|_, (x, y, z)| Ok(Value::Vector(x, y, z)))?,
    )?;
```

**Reply by setzer22** [reply]
Oh! I see. Thanks for the clarification :smile:

**Reply by setzer22** [reply]
That worked well. Not only that, but I got a 3x speedup over using tables 🎉

**Reply by khvzak** [reply]
Probably I'll register `vector(x,y,z)` function in mlua when creating Luau state to make it even faster (as it never fails).

**Reply by setzer22** [reply]
Ah, for sure! I'd be very interested in something like that.

Actually, is there a way for me as an mlua user to register these kind of infallible functions that are supposed to be called very frequently? I'm thinking of exposing things like noise functions and it would be nice if they had the least overhead possible.

**Reply by khvzak** [reply]
Generic way is use [`Lua::create_c_function`](https://docs.rs/mlua/0.8.0-beta.2/mlua/struct.Lua.html#method.create_c_function)
It creates a a Lua/C function with zero overhead. However the ffi module is not yet published so you need to add lua C functions manually (via extern "C" block). They are available in Lua manual.
I plan to publish ffi module as well for external use.

**Reply by khvzak** [reply]
@setzer22 you could try the new beta.3 with native `vector()` constructor support (it should be fast).

**Reply by setzer22** [reply]
Thanks! I made a quick test it and it seems to be working nicely :smile: Not a significant difference on my (completely improvised and not very scientific) benchmarks, but I guess that's to be expected with branch prediction. Still, nice to have the native constructor in :)
